homeassistant/components/energy/data.py

```diff
@@ -0,0 +1,265 @@
+    # an EnergyCostSensor will be automatically created
+    stat_cost: str | None
+
+    # Used to generate costs if stat_cost is set to None
+    entity_energy_from: str | None  # entity_id of an energy meter (kWh), entity_id of the energy meter for stat_energy_from
+    entity_energy_price: str | None  # entity_id of an entity providing price ($/kWh)
+    number_energy_price: float | None  # Price for energy ($/kWh)
+
+
+class FlowToGridSourceType(TypedDict):
+    """Dictionary describing the 'to' stat for the grid source."""
+
+    # kWh meter
+    stat_energy_to: str
+
+    # statistic_id of compensation ($) received for contributing back
+    # If set to None and entity_energy_from and entity_energy_price are configured,
+    # an EnergyCostSensor will be automatically created
+    stat_compensation: str | None
+
+    # Used to generate costs if stat_compensation is set to None
+    entity_energy_from: str | None  # entity_id of an energy meter (kWh), entity_id of the energy meter for stat_energy_from
+    entity_energy_price: str | None  # entity_id of an entity providing price ($/kWh)
+    number_energy_price: float | None  # Price for energy ($/kWh)
+
+
+class GridSourceType(TypedDict):
+    """Dictionary holding the source of grid energy consumption."""
+
+    type: Literal["grid"]
+
+    flow_from: list[FlowFromGridSourceType]
+    flow_to: list[FlowToGridSourceType]
+
+    cost_adjustment_day: float
+
+
+class SolarSourceType(TypedDict):
+    """Dictionary holding the source of energy production."""
+
+    type: Literal["solar"]
+
+    stat_energy_from: str
+    stat_predicted_energy_from: str | None
+
+
+SourceType = Union[GridSourceType, SolarSourceType]
+
+
+class DeviceConsumption(TypedDict):
+    """Dictionary holding the source of individual device consumption."""
+
+    # This is an ever increasing value
+    stat_consumption: str
+
+
+class EnergyPreferences(TypedDict):
+    """Dictionary holding the energy data."""
+
+    currency: str
+    energy_sources: list[SourceType]
+    device_consumption: list[DeviceConsumption]
+
+
+class EnergyPreferencesUpdate(EnergyPreferences, total=False):
+    """all types optional."""
+
+
+def _flow_from_ensure_single_price(
+    val: FlowFromGridSourceType,
+) -> FlowFromGridSourceType:
+    """Ensure we use a single price source."""
+    if (
+        val["entity_energy_price"] is not None
+        and val["number_energy_price"] is not None
+    ):
+        raise vol.Invalid("Define either an entity or a fixed number for the price")
+
+    return val
+
+
+FLOW_FROM_GRID_SOURCE_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Required("stat_energy_from"): str,
+            vol.Required("stat_cost"): vol.Any(None, str),
+            vol.Required("entity_energy_from"): vol.Any(None, str),
+            vol.Required("entity_energy_price"): vol.Any(None, str),
+            vol.Required("number_energy_price"): vol.Any(None, vol.Coerce(float)),
```

Comment:
Shouldn't these all be optional? As they call be `None`
```suggestion
            vol.Required("stat_cost"): vol.Any(None, str),
            vol.Required("entity_energy_from"): vol.Any(None, str),
            vol.Required("entity_energy_price"): vol.Any(None, str),
            vol.Required("number_energy_price"): vol.Any(None, vol.Coerce(float)),
```